### ISSUE TITLE
Convert Exploration Missions to new Mission Project approach

### ISSUE THREAD

**Issue by bevans2000** [mars-sim/mars-sim#975]
**Is your feature request related to a problem? Please describe.**
The Mission Project approach has been used to create an example TestDriveMission. This can be expanded to the EVA based Missions and using the building block already created.
This needs a new MissionProjectStep to perform the EVA step.

**Comment by bevans2000** [comment]
Needs #1324 to make loading easier and support old and new Missions

**Comment by mokun** [comment]
@bevans2000 

I'd say switching to the new mission project is one of the top priority for v3.9.0.

**Comment by bevans2000** [comment]
I think that also. So let's have that as a primary and solving the concurrent access problems. The latter will probably need other supporting changes.

**Comment by mokun** [comment]
So. How would it affect construction types of missions ?

**Comment by bevans2000** [comment]
> So. How would it affect construction types of missions ?

To be honest, I'm ot sure yet. But Missions are rigid at the moment and this will make them flexible. For the UI, the user will be offered options on the steps that can be added, so they essentially select what is done in what order.
Automatically,I think it's about the templates. So a Mission draft based on a template is scored. In theory all the logic exists, it just a case of calling it differently.

**Comment by mokun** [comment]
> There should be no subclasses of the generic mission because they all the same

Does this still your approach ?

Let's proceed with this if you think this is better.

Sounds like it will be a big change.

I think we could try implement this new mission project incrementally and in a parallel way.

So we could set aside the old mission codes in our code base for now.

All the new mission that we have envision so far will adopt the new paradigm that you have mentioned.

We probably need to create just a little bit of interfacing mechanism with the old code just to make the new mission approach work well with the old.

As soon as the new mission project work the way we want, we could start converting each existing mission one by one to the new.

How do you like that ?

**Comment by bevans2000** [comment]
Yes this will be better as it's more separated design which means we can unit test.
Let's go slow so next I'll create the historical visit mission using the new approach.

**Comment by mokun** [comment]
Currently, a person can get 'trapped' during the new Exploration project mission by being directly assigned to "ExploreSite" task non-stop. 

- The Exploration 32-20 did not have the Ended timestamp.
<img width="1440" height="771" alt="Image" src="https://github.com/user-attachments/assets/e45e3956-cf49-4753-8edb-390ec9b3579d" />

- The Exploration 32-20 has been "Closedown" though. It even said it was "Completed" in Phase Log when it's NOT.
<img width="550" height="529" alt="Image" src="https://github.com/user-attachments/assets/880aba09-1370-4564-8ff5-fcd791a5c633" />

- Even though he has arrived back at Star Base, he was left on MarsSurface still doing ExploreSite task.  
<img width="1731" height="469" alt="Image" src="https://github.com/user-attachments/assets/21aa0c27-ee4c-42e4-ab95-db03b5b997b0" />

- He was still directly assigned.
<img width="541" height="788" alt="Image" src="https://github.com/user-attachments/assets/0f5dca9a-19c8-42e2-b257-43af6434d5c8" />

- He's about to die.
<img width="540" height="828" alt="Image" src="https://github.com/user-attachments/assets/6f188509-28a3-492b-abf3-fad6c60898b8" />

**Comment by mokun** [comment]
Note that AbstractMission's endMission() did not currently process your brand new 'Closedown' phase yet.

I'm not sure if your new mission paradigm called it prematurely. It said the rover is still unloading.

It shouldn't be calling endMission() to complete the mission yet, right ?

```java
	/**
	 * Finalizes the mission. Reason for ending mission. Mission can
	 * override this to perform necessary finalizing operations.
	 *
	 * @param endStatus A status to add for the end of Mission
	 */
	protected void endMission(MissionStatus endStatus) {
		if (done) {
			logger.warning(startingMember, getName() + " already ended.");
			return;
		}

		// Ended with a status
		if (endStatus != null) {
			missionStatus.add(endStatus);
		}

		// If no mission flags have been added then it was accomplished
		String listOfStatuses = missionStatus.stream().map(MissionStatus::getName).collect(Collectors.joining(", "));
		MissionPhase finalPhase = ABORTED_PHASE;

		 if (aborted) {
			finalPhase = ABORTED_PHASE;
		}
		else if (endStatus == MISSION_ABORTED_BY_PLAYER) {
			finalPhase = ABORTED_PHASE;
		}
		else if (endStatus == MISSION_ACCOMPLISHED) {
			addMissionScore();
			finalPhase = COMPLETED_PHASE;
		}
		else {
			finalPhase = INCOMPLETE_PHASE;
		}
		
		setPhase(finalPhase, listOfStatuses);
		log.setDone();
		done = true;

		getAssociatedSettlement().getMissionControl().finishMission(!aborted);
		
		StringBuilder status = new StringBuilder();
		
		if (listOfStatuses.isBlank()) {
			status.append("Ended the ")
			.append(getName())
			.append(" without status flag(s).");
		}
		else {
			status.append("Ended the ")
			.append(getName())
			.append(" with status flag(s): ")
			.append(listOfStatuses).append(".");
		}

		logger.info(startingMember, status.toString());

		// Disband the members
		if (members != null && !members.isEmpty()) {
			String listOfMembers = members.stream().map(Worker::getName).collect(Collectors.joining(", "));
			logger.info(startingMember, DISBANDING + getFullMissionDesignation() + MEMBERS_ + listOfMembers);
			
			removeAllMembers();
		}

		setPhaseEnded(true);
		
		if (this instanceof AbstractVehicleMission avm) {
			avm.releaseVehicle();
		}
		else {
			endMission(MISSION_ABORTED_BY_PLAYER);
		}
		
		fireMissionUpdate(END_MISSION_EVENT);
	}
```

**Comment by mokun** [comment]
I was surprised to know that Christopher Cooper actually got ejected from Exploration 32-20.  

But somehow he was still collecting rocks later. In the Exploration site, or at the settlement vicinity ? 

```
01-Bora-04:371.532(Info) MissionBoardVehicleStep : [Star Base] Osprey - Look to evict members for Exploration 32-20
01-Bora-04:371.532(Info) MissionBoardVehicleStep : [Star Base] Christopher Cooper - Ejected from mission Exploration 32-20 missed Departure
01-Bora-04:371.532(Info) MissionBoardVehicleStep : [Star Base] Osprey - Ready to depart for Exploration 32-20
01-Bora-04:371.532(Info) Project : Exploration 32-20 started step Mission Exploration 32-20 travel to Mineral Site 29
...
01-Bora-04:445.615(Info) ExploreSite : [39.8000 N 157.9000 W] Christopher Cooper - Expected to collect 78.8 kg rocks to a specimen box.
...


```

**Comment by bevans2000** [comment]
> Note that AbstractMission's endMission() did not currently process your brand new 'Closedown' phase yet.
> 
Doesn't matter as the new approach does not use Abstract Mission it's a mess.

**Comment by bevans2000** [comment]
OK, looks like the Mission think he was ejected but the Person is still assigned the mission. This has nothing to do with directly assigned Task. The problem is earlier in the ejection. I'll update the UnitTest to trap it. This has nothing to do with Exploration but the common step.

**Comment by mokun** [comment]
> This has nothing to do with directly assigned Task.

Maybe. But isn't it that using a meta task for ExploreSite would be a lot better and safer in future ? 

There are directly assigned Walk task that would still pop up from time to time that would cause a person stop functioning normally.

**Comment by bevans2000** [comment]
No, it's the unstable logic within the Task that is the problem. The pattern of directly assigning tasks in the mission is save because the worker is looking for a new task so can accept it. They are not doing anything, that is why they are in the mission execute method.
Directly assigning when a Task is already active does cause problems. This has been increased with a worker being forced to eat or sleep from random code. That does cause problems because the original Task has to be stopped midflow and can leave an inconsistent state.
The exception is the Sleep Task but that is stopped via an alarm concept so it ends itself.

**Comment by mokun** [comment]
> This has been increased with a worker being forced to eat or sleep from random code. 

It's not exactly a random thing. 

I see many times a person was very tired and hungry and thirst and would still go out to do EVA, only to find out after a short while he would have to come back to the vehicle and rest and eat.

Before calling assignTask(new EatDrink()) or assignTask(new Sleep), it also checks how hungry and thirsty and tired, prior to exiting the vehicle to do EVA tasks.

If we can't make a person *willingly* choose to take care of his need of eating and drinking *prior to EVA*, it would be the best. 

We could probably capture these motivation in the logics within EatDrinkMeta.

If we use Meta task to compute what to do for Sleep and EatDrink, we will need to do it for all mission and non-mission related tasks such as ExploreSite and CollectResources across board. 

That's why I need your buy-in so that we can do it.

> The exception is the Sleep Task but that is stopped via an alarm concept so it ends itself

Yes for Sleep. But it's not the only exception. The EatDrink task has a total of 7 places calling for endTask() (as well as 2 calls to clearTask()) when conditions are met or not met.

```java
/**
	 * Performs eating preserved food phase.
	 *
	 * @param time
	 * @return
	 */
	private double eatingPreservedFoodPhase(double time) {
		
		// Checks if this person has eaten too much already 
		if (pc.eatTooMuch()) {
			endTask();
			return time;
		}
		
		// Give a person the chance to eat even if in high fatigue
		int rand = RandomUtil.getRandomInt(1);
		if (rand == 1 && pc.getFatigue() > 2_000) {
			endTask();
			return time;
		}
		
		double remainingTime = 0;
		double eatingTime = time;

		double proportion = eatPreservedFood(eatingTime);

		// FUTURE : If not enough preserved food available, 
		// should change to dessert phase.
		
		if (proportion == 0.0) {
			endTask();
			return time;
		}
		else {
```

**Comment by bevans2000** [comment]
I disagree and this is an example of having logic spread around the code base.
When a person calls mission execute they are in a state of not having anything to do so can be directly assigned. This has check that a person is fit for a mission of course.
Those Tasks do not need a meta task.
We should be simulating persons having a freewill so if there is no mission, then they should be free to choose there next Task within the scope of their Shift. The problem I see is the scores of the metatasks are imbalanced so someone very hungry is not getting a high enough score. There need to be choice based on score, for example if there is a major repair needed then people should forego sleep to fix it.
I would go further and enforce a task cannot be prematurely closed, Task cannot end them sleeps. Then a new top level Task can not be assigned unless there is not one present. Maybe some Tasks, e.g Sleep, need an interface to allow them to support a request to stop.

**Comment by mokun** [comment]
Currently, there are 3 places that calls checkNeeds(), 
1. (x2) RoverMission's creatOperateVehicleTask()
2. DroneMission's creatOperateVehicleTask()
3. CollectResourcesMission's performEVA()

In AbstractVehicleMission, we have

```java
	/**
	 * Check needs.
	 * 
	 * @param person
	 * @return
	 */
	public boolean checkNeeds(Person person) {
		
		PhysicalCondition pc = person.getPhysicalCondition();
		var persRH = person.getEquipmentInventory();

		// Note: if a person is not in fatigue but is hungry or thirsty, don't need to sleep
		double fatigue = person.getPhysicalCondition().getFatigue();
//		if (person.getPhysicalCondition().getFatigueLevel().getMaxValue() >= FatigueLevel.RESTED.getMaxValue()) {
		if (fatigue > 1000 && person.isInside()) {				
			boolean canSleep = assignTask(person, new Sleep(person, 100));
        	if (canSleep) {
        		logger.log(person, Level.INFO, 4_000,
            			"Instructed to sleep before piloting " + getVehicle() + ".  Fatigue: " + Math.round(fatigue) + ".");
        		
        		return false;
        	}
        	else {
				logger.warning(person, 4_000, "Unable to sleep as assigned.");
        	}
    	}
		
		else if (RandomUtil.getRandomInt(1) == 0 && (pc.isDoubleHungry())) {
			
			// Identify the available amount first
			double foodAmount = persRH.getSpecificAmountResourceStored(ResourceUtil.FOOD_ID);

			var localrh = EquipmentOwner.getAttached(person.getContainerUnit());
			if (localrh != null) {
				if (foodAmount < EatDrink.MIN)
					foodAmount = localrh.getSpecificAmountResourceStored(ResourceUtil.FOOD_ID);
			}
			
			if (foodAmount > EatDrink.MIN) {
				boolean canEatDrink = assignTask(person, new EatDrink(person, getVehicle()));
	        	if (canEatDrink) {
	        		logger.log(person, Level.INFO, 4_000,
	            			"Instructed to eat before piloting " + getVehicle() 
	            			+ ".  Hunger: " + Math.round(pc.getHunger()) + ".");
	        		
	        		return false;
	        	}
				else {
					logger.warning(person, 10_000, "Unable to eat as assigned.");
				}
			}
			else {
				logger.warning(person, 10_000, "No food left.");
			}
    	}	
		
		else if (RandomUtil.getRandomInt(1) == 0 && (pc.isDoubleThirsty())) {		
			
			// Identify the available amount first
			double waterAmount = persRH.getSpecificAmountResourceStored(ResourceUtil.WATER_ID);
			
			var localrh = EquipmentOwner.getAttached(person.getContainerUnit());
			if (localrh != null) {
				if (waterAmount < EatDrink.MIN)
					waterAmount = localrh.getSpecificAmountResourceStored(ResourceUtil.WATER_ID);
			}
			
			if (waterAmount > EatDrink.MIN) {
				boolean canEatDrink = assignTask(person, new EatDrink(person, getVehicle()));
	        	if (canEatDrink) {
	        		logger.log(person, Level.INFO, 4_000,
	            			"Instructed to drink before piloting " + getVehicle() 
	            			+ ".  Thirst: " + Math.round(pc.getThirst()) + ".");
	        		
	        		return false;
	        	}
				else {
					logger.warning(person, 10_000, "Unable to drink as assigned.");
				}
			}
			else {
				logger.warning(person, 10_000, "No water left.");
			}
    	}	
		
		return true;
	}

```

I would love to make eating, drinking and sleeping being completely voluntary tasks and not being forced upon.

By the way, eating and drinking can be just subtasks. Can't they ?

**Comment by bevans2000** [comment]
With the new mission approach I want a clean start and code whatnwe need, not what we have already which is complex. I've got more tests to add
